fosrl-pangolin: 1.2.0 -> 1.8.0

SigmaSquadron opened the pull request:
Changelogs: [1.3.0](https://github.com/fosrl/pangolin/releases/tag/1.3.0), [1.3.1](https://github.com/fosrl/pangolin/releases/tag/1.3.1), [1.3.2](https://github.com/fosrl/pangolin/releases/tag/1.3.2), [1.4.0](https://github.com/fosrl/pangolin/releases/tag/1.4.0), [1.5.0](https://github.com/fosrl/pangolin/releases/tag/1.5.0), [1.5.1](https://github.com/fosrl/pangolin/releases/tag/1.5.1), [1.6.0](https://github.com/fosrl/pangolin/releases/tag/1.6.0), [1.6.1](https://github.com/fosrl/pangolin/releases/tag/1.6.1), [1.6.2](https://github.com/fosrl/pangolin/releases/tag/1.6.2), [1.7.0](https://github.com/fosrl/pangolin/releases/tag/1.7.0), [1.7.1](https://github.com/fosrl/pangolin/releases/tag/1.7.1), [1.7.2](https://github.com/fosrl/pangolin/releases/tag/1.7.2), [1.7.3](https://github.com/fosrl/pangolin/releases/tag/1.7.3), [1.8.0](https://github.com/fosrl/pangolin/releases/tag/1.8.0).

This combined update introduces a new database type for Pangolin: PostgreSQL. The module for it will come at a later date, so no separate package for it right now.

1.6.0 also introduces the `pangctl` command, which has inspired me to write the spaghetti you see in the `preFixup` section of the diff. By wrapping the `*.mjs` files to actual commands in `$out/bin`, we can simplify the NixOS module, make it easier to access `pangctl`, and allow non-NixOS users to start a Pangolin server without much hassle.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jackrosenberg reviewed (approved):
Lgtm

SigmaSquadron commented:
Updated to 1.6.2. The highlight of this update is the new commands in `$out/bin`. I've also taken the opportunity to move the Pangolin scripts and static files to `$out/share/pangolin` so we can have some resemblance of FHS paths instead of the mess NextJS forces unto us.

jackrosenberg commented on pkgs/by-name/fo/fosrl-pangolin/package.nix:
```diff
@@ -35,32 +64,104 @@ buildNpmPackage rec {
+
+  preInstall = ''
+    mkdir -p $out/{bin,share/pangolin}
   '';
```
is there a reason this is not in the `installPhase`?

SigmaSquadron replied:
I wanted to separate the phases as much as possible to give people overriding pangolin as many options as possible.

SigmaSquadron committed:
fosrl-pangolin: 1.2.0 -> 1.8.0

- This update introduces the PostgreSQL database type, which can be
  selected instead of SQLite.
- The derivation is now using `finalAttrs`, as buildNpmPackage supports
  it, so there's no reason to use `rec`.
- Finally, in order to simplify the module and allow non-NixOS users to
  use the Pangolin server, the minified `.mjs` files have been wrapped
  into shell scripts that can set up an appropriate environment for the
  server.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

jackrosenberg reviewed (changes requested):
Revoked approval since nixos/pangolin needs to be updated at the same time as updating to 1.6 and above.

jackrosenberg commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 419533`
Commit: `d299df66b63ba4e6f2f98f5b1e13bedd7b748d8b`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>fosrl-pangolin</li>
  </ul>
</details>

ethancedwards8 commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 419533`
Commit: `d299df66b63ba4e6f2f98f5b1e13bedd7b748d8b`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>fosrl-pangolin</li>
  </ul>
</details>

NickCao commented on pkgs/by-name/fo/fosrl-pangolin/package.nix:
```diff
@@ -35,32 +64,99 @@ buildNpmPackage rec {
   '';
 
+  postInstall = ''
+    sed -i '1i #! /usr/bin/env node' $out/share/pangolin/dist/{migrations,server}.mjs
```
This would be fixed up by patchShebangs or not?

SigmaSquadron replied:
It wouldn't. This is adding a shebang for `patchShebangs` to fix.

I guess this can be fixed upstream though.

SigmaSquadron replied:
Opened https://github.com/fosrl/pangolin/pull/1139

SigmaSquadron replied:
This was fixed upstream, so the `postInstall` was removed.

NickCao commented on pkgs/by-name/fo/fosrl-pangolin/package.nix:
```diff
@@ -35,32 +64,99 @@ buildNpmPackage rec {
+      (
+        attr:
+        "makeWrapper $out/share/pangolin/dist/${attr.mjs}.mjs $out/bin/${attr.command} ${
+          variablesMapped (attr.mjs == "server")
```
Why's constructing the command line flags done here not in the nixos module?

SigmaSquadron replied:
Should be done now.


<sub><sup>Don't ask why the commit is suddenly a cherry-pick.<sup><sub>

jackrosenberg commented:
@NickCao would you mind taking another look at this?

NickCao commented:
Still not very convinced by the make it work out of nixos part (since most of these users would happily use docker instead), but otherwise lgtm.

SigmaSquadron commented:
Well, this also has the benefit of making the module simpler. The `ollama-nextjs-webui` package also has something similar.
